From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Sat, 4 Apr 2026 20:03:09 -0400
Subject: [PATCH] Paper: Lazy set Entity.projectileSource in
 AbstractProjectile#getShooter

Original license: GPL-3.0-only
Original project: https://github.com/PaperMC/Paper
Paper commit: https://github.com/PaperMC/Paper/commit/5e9a9188c04668e8f6b9ebda1060f295ca9b4baa

Backported from Paper 26.1.1

Fixes Paper#13475
Supersedes Paper#13643
Supersedes Paper#13481

diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index b1c22ae1a6ab3f25abc851c145fb25586099ffe1..20ee8c7c8bf28d6f1cd1ed540c5fb4c3a04ba550 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -102,25 +102,12 @@ public abstract class Projectile extends Entity implements TraceableEntity {
 
     protected void setOwner(@Nullable EntityReference<Entity> owner) {
         this.owner = owner;
-        this.refreshProjectileSource(false); // Paper
     }
 
     public void setOwner(@Nullable Entity owner) {
         this.setOwner(EntityReference.of(owner));
     }
 
-    // Paper start - Refresh ProjectileSource for projectiles
-    public void refreshProjectileSource(boolean fillCache) {
-        if (fillCache) {
-            this.getOwner();
-        }
-        Entity owner = this.getOwner();
-        if (owner != null && this.projectileSource == null && owner.getBukkitEntity() instanceof org.bukkit.projectiles.ProjectileSource projSource) {
-            this.projectileSource = projSource;
-        }
-    }
-    // Paper end - Refresh ProjectileSource for projectiles
-
     @Override
     public @Nullable Entity getOwner() {
         return EntityReference.getEntity(this.owner, this.level());
